src/unitxt/operators.py

```diff
@@ -508,75 +508,124 @@ def process_value(self, value: Any) -> Any:
+            else:
+                for k in arg:
+                    assert isinstance(
+                        k, str
+                    ), f"{arg_name} should be a list of strings, whereas member {k!s} is of type {type(k)}"
+        assert (
+            self.prefix_len > 0
+        ), f"prefix_len must be positive, got {self.prefix_len}"
+        assert (
+            self.suffix_len > 0
+        ), f"suffix_len must be positive, got {self.suffix_len}"
         super().verify()
 
-    def prepare(self):
-        self.pats = (
-            self.suffixes
-            if isinstance(self.suffixes, list)
-            else [k for k, v in self.suffixes.items()]
+    def _calculate_distribution(self, end, end_name):
+        if end is None:
+            return
+        self.pats[end_name] = (
+            end if isinstance(end, list) else [k for k, v in end.items()]
         )
         total_weight = (
-            len(self.pats)
-            if isinstance(self.suffixes, list)
-            else sum([v for k, v in self.suffixes.items()])
+            len(self.pats[end_name])
+            if isinstance(end, list)
+            else sum([v for k, v in end.items()])
         )
-        self.weights = (
-            [1.0 / total_weight] * len(self.pats)
-            if isinstance(self.suffixes, list)
-            else [float(self.suffixes[p]) / total_weight for p in self.pats]
+        self.weights[end_name] = (
+            [1.0 / total_weight] * len(self.pats[end_name])
+            if isinstance(end, list)
+            else [float(end[p]) / total_weight for p in self.pats[end_name]]
         )
 
+    def prepare(self):
+        # Being an artifact, prepare is invoked before verify. Here we need verify before the actions
+        self.verify()
+        self.pats = {"prefixes": None, "suffixes": None}
+        self.weights = {"prefixes": None, "suffixes": None}
+        self.lens = {"prefixes": self.prefix_len, "suffixes": self.suffix_len}
+
+        self._calculate_distribution(self.prefixes, "prefixes")
+        self._calculate_distribution(self.suffixes, "suffixes")
         super().prepare()
 
+    def _get_random_pattern(self, end_name: str) -> str:
+        string_to_add = ""
+        if self.pats[end_name]:
+            string_to_add = "".join(
+                get_random().choices(self.pats[end_name], self.weights[end_name], k=1)
+                * self.lens[end_name]
+            )
+        return string_to_add
+
     def process_value(self, value: Any) -> Any:
         assert value is not None, "input value should not be None"
         new_value = str(value)
-        if self.remove_existing_trailing_whitespaces:
-            new_value = new_value.rstrip()
-        new_value += get_random().choices(self.pats, self.weights, k=1)[0]
-
-        return new_value
+        if self.remove_existing_whitespaces:
+            new_value = new_value.strip()
+        prefix = self._get_random_pattern("prefixes")
```

Comment:
Here also - best to pass in  pats["suffixes"], and weights["suffixes"] and not extract it in the functions.

Perhaps also call pats -> patterns

Reply:
Thanks!  Done.